dbmgr/dbmgr.go

```diff
@@ -0,0 +1,985 @@
+	db.AutoMigrate(&ObjRef{})
+	db.AutoMigrate(&Collection{})
+	db.AutoMigrate(&CollectionRef{})
+
+	db.AutoMigrate(&Deal{})
+	db.AutoMigrate(&DFERecord{})
+	db.AutoMigrate(&PieceCommRecord{})
+	db.AutoMigrate(&ProposalRecord{})
+	db.AutoMigrate(&RetrievalFailureRecord{})
+	db.AutoMigrate(&RetrievalSuccessRecord{})
+
+	db.AutoMigrate(&MinerStorageAsk{})
+	db.AutoMigrate(&StorageMiner{})
+
+	db.AutoMigrate(&User{})
+	db.AutoMigrate(&AuthToken{})
+	db.AutoMigrate(&InviteCode{})
+
+	db.AutoMigrate(&Shuttle{})
+
+	var count int64
+	if err := db.Model(&StorageMiner{}).Count(&count).Error; err != nil {
+		return nil, err
+	}
+
+	//miners from minerX spreadsheet
+	minerStrs := []string{
+		"f02620",
+		"f023971",
+		"f022142",
+		"f019551",
+		"f01240",
+		"f01247",
+		"f01278",
+		"f071624",
+		"f0135078",
+		"f022352",
+		"f014768",
+		"f022163",
+		"f09848",
+		"f02576",
+		"f02606",
+		"f019041",
+		"f010617",
+		"f023467",
+		"f01276",
+		"f02401",
+		"f02387",
+		"f019104",
+		"f099608",
+		"f062353",
+		"f07998",
+		"f019362",
+		"f019100",
+		"f014409",
+		"f066596",
+		"f01234",
+		"f058369",
+		"f08399",
+		"f021716",
+		"f010479",
+		"f08403",
+		"f01277",
+		"f015927",
+	}
+
+	var defaultMiners []address.Address
+	for _, s := range minerStrs {
+		a, err := address.NewFromString(s)
+		if err != nil {
+			panic(err)
+		}
+
+		defaultMiners = append(defaultMiners, a)
+	}
+
+	if count == 0 {
+		// TODO: this could go into its own generic function, potentially batch
+		// these insertions
+		fmt.Println("adding default miner list to database...")
+		for _, m := range defaultMiners {
+			db.Create(&StorageMiner{Address: util.DbAddr{Addr: m}})
+		}
+
+	}
+
+	return &DBMgr{db}, nil
+}
+
+// USERS
+
+type UsersQuery struct{ DB *gorm.DB }
+type UserID uint
+type User struct {
+	gorm.Model
+
+	UUID     string `gorm:"unique"`
+	Username string `gorm:"unique"`
+	PassHash string
+
+	UserEmail string
+
+	Address util.DbAddr
+
+	Perm  int
+	Flags int
+
+	StorageDisabled bool
+}
+
+func NewUsersQuery(db *gorm.DB) *UsersQuery {
+	return &UsersQuery{DB: db.Model(User{})}
+}
+
+func (q *UsersQuery) WithUsername(username string) *UsersQuery {
+	q.DB = q.DB.Where("username = ?", username)
+	return q
+}
+
+func (q *UsersQuery) WithID(id uint) *UsersQuery {
+	q.DB = q.DB.Where("id = ?", id)
+	return q
+}
+
+func (q *UsersQuery) Create(user User) error {
+	return q.DB.Create(&user).Error
+}
+
+func (q *UsersQuery) Get() (User, error) {
+	var user User
+	if err := q.DB.Take(&user).Error; err != nil {
+		return User{}, err
+	}
+	return user, nil
+}
+
+func (q *UsersQuery) Count() (int64, error) {
+	var count int64
+	if err := q.DB.Count(&count).Error; err != nil {
+		return 0, err
+	}
+	return count, nil
+}
+
+func (q *UsersQuery) Exists() (bool, error) {
+	count, err := q.Count()
+	if err != nil {
+		return false, err
+	}
+	return count > 0, nil
+}
+
+// Errors if none were deleted
+func (q *UsersQuery) ExpectDelete() error {
+	res := q.DB.Delete(&User{})
```

Comment:
it might be worth throwing some safety checks in here to make sure nobody just calls this by itself. I think that would delete all users...

Reply:
I was thinking about that. My idea was to put a boolean in relevant query struct isSingle or something along those lines. Queries that fundamentally select one item would get the single flag turned on after being called, and certain actions could throw an error if they weren't called on a single query. 

Even nicer to use would be to have an additional struct, TableNameHereQuerySingle, which would embed TableNameHereQuery and would expose additional functions that should only be called for single queries and not multi line queries. That would make issues show up at compile time. It adds a bit more complexity though, so might want to wait for generics for this approach. I think it sounds super cool though.

Reply:
Yeah, i think just adding a flag that says whether *any* condition has been set would be good